### PR TITLE
Adds an experimental store/load add to config bag

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
- store multiple objects of the same type in the bag
- obtain mutable access to items in the bag

## Description
This adds `store_append` to enable storing multiple items of the same type in the bag and `get_mut`, `get_mut_or_else` and `get_mut_or_default` to reduce clones when possible

## Testing
UTs

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: rust-runtime/aws-smithy-runtime-api/src/config_bag.rs [hunk 1]
```diff
@@ -59,42 +61,196 @@ impl Deref for FrozenConfigBag {
     }
 }
 
-pub trait Persist {
-    fn layer_name(&self) -> &'static str;
-    fn persist(&self, layer: &mut ConfigBag);
+pub(crate) mod value {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime-api/src/config_bag.rs:64]: What is the purpose of the module?

File: rust-runtime/aws-smithy-runtime-api/src/config_bag.rs [hunk 2]
```diff
@@ -310,55 +544,45 @@ mod test {
...
+        assert_eq!(open_bag.layers().count(), 4);
     }
 
     #[test]
-    fn persist_trait() {
-        #[derive(Debug, Eq, PartialEq, Clone)]
-        struct MyConfig {
-            a: bool,
-            b: String,
+    fn store_append() {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime-api/src/config_bag.rs:560]: Should add a test for iteration if there isn't one already

**Review by jdisanti — APPROVED** [review]
Looks good! Couple of minor things

File: rust-runtime/aws-smithy-runtime-api/src/config_bag.rs (hunk 1 above)
Comment (rcoh) [reply rust-runtime/aws-smithy-runtime-api/src/config_bag.rs:64]: keep value private while allowing it to be exposed in the trait

Comment (jdisanti) [reply rust-runtime/aws-smithy-runtime-api/src/config_bag.rs:64]: Can you add a comment?

**Review by jdisanti — APPROVED** [review]
LGTM!

**Pull request merged**